src/lib/expression/parameter_expression.cpp

```diff
@@ -0,0 +1,132 @@
+  }
+  stream << "id=" << std::to_string(parameter_id);
+
+  stream << "]";
+
+  if (_value) {
+    stream << "=" << *_value;
+  }
+
+  return stream.str();
+}
+
+DataType ParameterExpression::data_type() const {
+  if (parameter_expression_type == ParameterExpressionType::ValuePlaceholder) {
+    Assert(_value.has_value(), "Can't obtain type of unset ValuePlaceholder");
+    return data_type_from_all_type_variant(*_value);
+  } else {
+    return _referenced_expression_info->data_type;
+  }
+}
+
+bool ParameterExpression::is_nullable() const {
+  if (parameter_expression_type == ParameterExpressionType::ValuePlaceholder) {
+    Assert(_value.has_value(), "Can't obtain nullability of unset ValuePlaceholder");
+    return variant_is_null(*_value);
+  } else {
+    return _referenced_expression_info->nullable;
+  }
+}
+
+const std::optional<AllTypeVariant>& ParameterExpression::value() const { return _value; }
+
+void ParameterExpression::set_value(const std::optional<AllTypeVariant>& value) {
+  if (!value) {
+    _value.reset();
+  } else {
+    Assert(parameter_expression_type == ParameterExpressionType::ValuePlaceholder ||
+               data_type_from_all_type_variant(*value) == _referenced_expression_info->data_type,
+           "Can't set Parameter to this DataType");
+    _value = value;
+  }
+}
+
+bool ParameterExpression::_shallow_equals(const AbstractExpression& expression) const {
+  const auto& parameter_expression_rhs = static_cast<const ParameterExpression&>(expression);
+
+  // For the purpose of comparing ParameterExpressions, Null==Null
+  const auto both_are_null = _value && parameter_expression_rhs._value &&
+                             variant_is_null(*parameter_expression_rhs._value) && variant_is_null(*_value);
+
+  return parameter_id == parameter_expression_rhs.parameter_id &&
+         parameter_expression_type == parameter_expression_rhs.parameter_expression_type &&
+         (both_are_null || _value == parameter_expression_rhs._value) &&
+         _referenced_expression_info == parameter_expression_rhs._referenced_expression_info;
+}
+
+size_t ParameterExpression::_on_hash() const {
+  auto hash = boost::hash_value(parameter_id.t);
+
+  boost::hash_combine(hash, static_cast<std::underlying_type_t<ParameterExpressionType>>(parameter_expression_type));
+  boost::hash_combine(hash, std::hash<AllTypeVariant>{}(_value.has_value()));
+  boost::hash_combine(hash, std::hash<AllTypeVariant>{}(_value.value_or(NullValue{})));
+  boost::hash_combine(hash, _referenced_expression_info.has_value());
+
+  if (_referenced_expression_info) {
+    boost::hash_combine(hash, static_cast<std::underlying_type_t<DataType>>(_referenced_expression_info->data_type));
+    boost::hash_combine(hash, _referenced_expression_info->nullable);
+    boost::hash_combine(hash, _referenced_expression_info->column_name);
+  } else {
+    boost::hash_combine(hash, 0);
+    boost::hash_combine(hash, 0);
+    boost::hash_combine(hash, 0);
```

Comment:
I thought this might avoid hash collisions if the parameter expression is nested in another expression, but it is probably just confusing so I removed the else branch.